plugins/omdtls/omdtls.c

```diff
@@ -0,0 +1,885 @@
+		(struct sockaddr*)&pWrkrData->dtls_client_addr,
+		sizeof(pWrkrData->dtls_client_addr)) < 0) {
+		LogError(0, RS_RET_SUSPENDED, "omdtls[%p]: dtls_init Unable to bind DTLS CLient socket with errno %d",
+			pWrkrData, errno);
+		ABORT_FINALIZE(RS_RET_COULD_NOT_BIND);
+	}
+
+	// Resolve target address
+	memset(&hints, 0, sizeof(hints));
+	/* port must be numeric, because config file syntax requires this */
+	hints.ai_family = glbl.GetDefPFFamily(runModConf->pConf);
+	hints.ai_socktype = SOCK_DGRAM;
+	if((iErr = (getaddrinfo((char*)pData->target, (char*)pData->port, &hints, &addrResolved))) != 0) {
+		LogError(0, RS_RET_SUSPENDED,
+			"omdtls[%p]: could not get addrinfo for hostname '%s':'%s': %s",
+			pWrkrData, pData->target, pData->port, gai_strerror(iErr));
+		ABORT_FINALIZE(RS_RET_ADDRESS_UNKNOWN);
+	}
+	pWrkrData->dtls_rcvr_addrinfo = addrResolved;
+	addrResolved = NULL;
+
+	// Init Socket Connection (Which technically does not connect but prepares socket for DTLS)
+	DBGPRINTF("dtls_init[%p]: Init Session to %s:%s\n", pWrkrData, pData->target, pData->port);
+
+	pWrkrData->sockout = socket(AF_INET, SOCK_DGRAM, 0);
+	// Connect the UDP socket to the server's address
+	//	(const struct sockaddr *) &pWrkrData->dtls_rcvr_addr, sizeof(pWrkrData->dtls_rcvr_addr)) 
+	if (connect(pWrkrData->sockout,
+		pWrkrData->dtls_rcvr_addrinfo->ai_addr,
+		pWrkrData->dtls_rcvr_addrinfo->ai_addrlen) < 0) {
+		LogError(0, RS_RET_SUSPENDED,
+			"dtls_init[%p]: Failed to connect to hostname '%s':'%s': %s",
+			pWrkrData, pData->target, pData->port, gai_strerror(iErr));
+		ABORT_FINALIZE(RS_RET_ERR);
+	}
+	
+	// Socket Connect successfull, no continue with TLS
+	CHKiRet(dtls_connect(pWrkrData));
+
+	// We reached this position means we are ready to send!
+	pWrkrData->bIsConnecting = 0;
+	pWrkrData->bIsConnected = 1;
+finalize_it:
+	DBGPRINTF("dtls_init[%p]: doTryResume %s iRet %d\n", pWrkrData, pData->target, iRet);
+	if(addrResolved != NULL) {
+		freeaddrinfo(addrResolved);
+	}
+	if(iRet != RS_RET_OK) {
+		if(pWrkrData->dtls_rcvr_addrinfo != NULL) {
+			freeaddrinfo(pWrkrData->dtls_rcvr_addrinfo);
+			pWrkrData->dtls_rcvr_addrinfo = NULL;
+		}
+		iRet = RS_RET_SUSPENDED;
+		pWrkrData->bIsConnected = 0;
+	}
+	RETiRet;
+}
+
+static rsRetVal
+dtls_close(wrkrInstanceData_t *pWrkrData)
+{
+	DEFiRet;
+	int iErr;
+	instanceData *const pData = (instanceData *const) pWrkrData->pData;
+	if (pWrkrData->bIsConnected) {
+		DBGPRINTF("dtls_close[%p]: close session for %s:%s\n", pWrkrData, pData->target, pData->port);
+		if (	pWrkrData->sslClient != NULL &&
+			pWrkrData->bIsConnected) {
+			iErr = SSL_shutdown(pWrkrData->sslClient);
+			if (iErr <= 0){
+				/* Shutdown not finished, call SSL_read to do a bidirectional shutdown, see doc:
+				*	https://www.openssl.org/docs/man1.1.1/man3/SSL_shutdown.html
+				*/
+				char rcvBuf[DTLS_MAX_RCVBUF];
+				SSL_read(pWrkrData->sslClient, rcvBuf, DTLS_MAX_RCVBUF);
+			}
+			SSL_free(pWrkrData->sslClient);
+			pWrkrData->sslClient = NULL;
```

Comment:
this looks like we need to loop and retry the shutdown?

Reply:
We do a single SSL_read which should be fine for protocol handling if necessary. We do not expect data to be received from the server.